HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,61 @@
+import findspark
+findspark.init()
+import pyspark
+
+
+def distance_to_all_nodes_spark(root_node, graph_rdd):
+    """A fast parallel implementation of the search algorithm
+       (does leave RDD for neighbors, for performace reasons in small graphs)."""
+    result = graph_rdd.context.parallelize([(root_node, 0)])
+    visited = {root_node}  # set initial node
+    neighbors = set(graph_rdd.lookup(root_node)[0])   # set initial neighbors
+    nodes = graph_rdd.context.accumulator(1)
+    iteration = 1
+    while neighbors:  # we are done when neigbors is empty
+        current = graph_rdd.filter(lambda (k, v): k in neighbors)
+        current = current.mapValues(lambda v: iteration)
+        new_neighbors = set(current.flatMap(lambda x: x).collect())
+        result = result.union(current)
+        nodes += len(neighbors)
+        visited = visited.union(neighbors)
+        neighbors = new_neighbors - visited
+        iteration += 1
+    print "Number of nodes: " + str(nodes.value)
+    return result
+
+
+def distance_to_all_nodes_edge(root_node, edges_rdd, N):
+    """Using edge tuples to determine distances to all nodes, without leaving Spark."""
+    edges_rdd = edges_rdd.partitionBy(N)
+    edges_rdd = edges_rdd.cache()
+    result = edges_rdd.context.parallelize([(root_node, 0)])
+    rdd = result
```

Comment:
A better name could be used for 'red', for instance, 'frontier' may be exactly the term in BFS.